pkg/objstore/oci/oci.go

```diff
@@ -0,0 +1,331 @@
+func (b *Bucket) Attributes(ctx context.Context, name string) (objstore.ObjectAttributes, error) {
+	response, err := getObject(ctx, *b, name, "")
+	if err != nil {
+		return objstore.ObjectAttributes{}, err
+	}
+	return objstore.ObjectAttributes{
+		Size:         *response.ContentLength,
+		LastModified: response.LastModified.Time,
+	}, nil
+}
+
+// createBucket creates bucket.
+func (b *Bucket) createBucket(ctx context.Context, compartmentId string) (err error) {
+	return createBucket(ctx, b, compartmentId)
+}
+
+// deleteBucket deletes bucket.
+func (b *Bucket) deleteBucket(ctx context.Context) (err error) {
+	return deleteBucket(ctx, b)
+}
+
+// NewBucket returns a new Bucket using the provided oci config values.
+func NewBucket(logger log.Logger, ociConfig []byte) (*Bucket, error) {
+	level.Debug(logger).Log("msg", "creating new oci bucket connection")
+	var config = DefaultConfig
+	var configurationProvider common.ConfigurationProvider
+	var err error
+
+	if err := yaml.Unmarshal(ociConfig, &config); err != nil {
+		return nil, errors.Wrapf(err, "unable to unmarshal the given oci configurations")
+	}
+
+	provider := Provider(strings.ToLower(config.Provider))
+	level.Info(logger).Log("msg", fmt.Sprintf("Creating OCI default '%s' provider", provider))
+	switch provider {
+	case DefaultConfigProvider:
+		configurationProvider = common.DefaultConfigProvider()
+	case InstancePrincipalConfigProvider:
+		configurationProvider, err = auth.InstancePrincipalConfigurationProvider()
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to create OCI instance principal config provider")
+		}
+	case RawConfigProvider:
+		if err := config.validateConfig(); err != nil {
+			return nil, errors.Wrapf(err, "invalid oci configurations")
+		}
+		configurationProvider = common.NewRawConfigurationProvider(config.Tenancy, config.User, config.Region,
+			config.Fingerprint, config.PrivateKey, &config.Passphrase)
+	default:
+		return nil, errors.Wrapf(err, fmt.Sprintf("unsupported OCI provider: %s", provider))
+	}
+
+	client, err := objectstorage.NewObjectStorageClientWithConfigurationProvider(configurationProvider)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to create ObjectStorage client with the given oci configurations")
+	}
+
+	httpClient := http.Client{
+		Transport: DefaultTransport(config),
+	}
+	client.HTTPClient = &httpClient
+
+	requestMetadata := getRequestMetadata(config.MaxRequestRetries, config.RequestRetryInterval)
+
+	namespace, err := getNamespace(client, requestMetadata)
+	if err != nil {
+		return nil, err
+	}
+	level.Debug(logger).Log("msg", fmt.Sprintf("OCI tenancy namespace: %s", *namespace))
+
+	bkt := Bucket{
+		logger:          logger,
+		name:            config.Bucket,
+		namespace:       *namespace,
+		client:          &client,
+		partSize:        config.PartSize,
+		requestMetadata: requestMetadata,
+	}
+	level.Info(logger).Log("msg", fmt.Sprintf("[Bucket] name: '%s'", bkt.name))
```

Comment:
Do we need to log this? (:

Reply:
Not really. Just my bad habit :)

Removed.